osu.Server.Queues.ScoreStatisticsProcessor/Processors/MedalAwarders/StarRatingMedalAwarder.cs

```diff
@@ -0,0 +1,154 @@
+﻿using JetBrains.Annotations;
+using MySqlConnector;
+using osu.Game.Online.API.Requests.Responses;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Difficulty;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Scoring;
+using osu.Server.Queues.ScoreStatisticsProcessor.Helpers;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+using osu.Server.Queues.ScoreStatisticsProcessor.Stores;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Processors.MedalAwarders
+{
+    [UsedImplicitly]
+    [SuppressMessage("ReSharper", "MergeIntoPattern")]
+    public class StarRatingMedalAwarder : IMedalAwarder
+    {
+        private BeatmapStore? beatmapStore;
+
+        public bool RunOnFailedScores => false;
+
+        public IEnumerable<Medal> Check(SoloScoreInfo score, UserStats userStats, IEnumerable<Medal> medals, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            return checkAsync(score, medals, conn, transaction).Result;
+        }
+
+        private async Task<IEnumerable<Medal>> checkAsync(SoloScoreInfo score, IEnumerable<Medal> medals, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            List<Medal> earnedMedals = new List<Medal>();
+
+            Ruleset ruleset = LegacyRulesetHelper.GetRulesetFromLegacyId(score.RulesetID);
+            Mod[] mods = score.Mods.Select(m => m.ToMod(ruleset)).ToArray();
+
+            // Ensure the score isn't using any difficulty reducing mods
+            if (mods.Any(x => x is ModEasy || x is ModNoFail || x is ModHalfTime))
+                return earnedMedals;
+
+            try
+            {
+                beatmapStore ??= await BeatmapStore.CreateAsync(conn, transaction);
+
+                Beatmap? beatmap = await beatmapStore.GetBeatmapAsync(score.BeatmapID, conn, transaction);
+                if (beatmap == null)
+                    return earnedMedals;
```

Comment:
Notably the old implementation excludes qualified and loved beatmaps here:

```php
if ($diffReduction || $beatmap['approved'] > 2) break;
```

Unsure whether that's something to preserve.

This applies to all of the medal ranges in this file.

Reply:
Added this, as especially Loved maps tend to sometimes have weird elements in them that break SR, so this is likely desired.

https://github.com/ppy/osu-queue-score-statistics/blob/d5f1eb7d096c083510a75951d52a0afec8858f4a/osu.Server.Queues.ScoreStatisticsProcessor/Processors/MedalAwarders/StarRatingMedalAwarder.cs#L56-L58